sphinx/themes/basic/static/searchtools.js

```diff
@@ -510,9 +511,26 @@ const Search = {
       if (word.length > 2) {
         const escapedWord = _escapeRegExp(word);
         if (!terms.hasOwnProperty(word)) {
-          Object.keys(terms).forEach((term) => {
-            if (term.match(escapedWord))
-              arr.push({ files: terms[term], score: Scorer.partialTerm });
+
+          let ngramTerms = function (ngram) {
+            let [node, path] = [termsNgrams, ""];
+            for (const step of ngram) {
+              if ((path += step) in node) [node, path] = [node[path], ""];
+            }
+            if (node.length === undefined) node = [node];
+            return node.map(offset => Object.keys(terms)[offset]);
+          };
+
+          const candidateTerms = new Set(ngramTerms(word.substring(0, 3)));
+          for (let start = 1; candidateTerms.size && start + 3 <= word.length; start++) {
+            const subsequentTerms = new Set(ngramTerms(word.substring(start, start + 3)));
+            for (const candidateTerm of candidateTerms) {
+              if (!subsequentTerms.has(candidateTerm)) candidateTerms.delete(candidateTerm);
+            }
+          }
```

Comment:
My guess is that this is the slowest part of the code client-side.  The performance traces I showed were for a string of length `4` (`phin` - Sphinx without the first and last letters) - for that this logic would perform two passes (checking `phi` and `hin`).

A possible alternative would be to begin with all terms matched for the first ngram, and then use brute-force on those terms -- it should be a much smaller set than the entire content of `terms`.

Reply:
The `ngramsTerms` lookup might also be expensive; I wonder whether the `Object.keys` call in there is costly.

Reply:
> The `ngramsTerms` lookup might also be expensive; I wonder whether the `Object.keys` call in there is costly.

:exclamation: - I think we're invoking that `Object.keys` call for every single term-offset retrieved.  So yep, that's worth checking and refactoring.